File: src/main/java/fi/metatavu/metaform/server/rest/translate/EmailNotificationTranslator.java
package fi.metatavu.metaform.server.rest.translate;

import java.io.IOException;
import org.slf4j.Logger;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;

import org.apache.commons.lang3.StringUtils;

import com.fasterxml.jackson.databind.ObjectMapper;

import fi.metatavu.metaform.server.notifications.EmailNotificationController;
import fi.metatavu.metaform.server.rest.model.EmailNotification;
import fi.metatavu.metaform.server.rest.model.FieldRule;

/**
 * Translator for email notifications
 * 
 * @author Antti Leppä
 * @author Heikki Kurhinen
 */
@ApplicationScoped
public class EmailNotificationTranslator {

  @Inject
  private Logger logger;
  
  @Inject
  private EmailNotificationController emailNotificationController;

  /**
   * Translates JPA email notification object into REST email notification object
   * 
   * @param emailNotification JPA emailNotification object
   * @return REST emailNotification
   */
  public EmailNotification translateEmailNotification(fi.metatavu.metaform.server.persistence.model.notifications.EmailNotification emailNotification) {
    if (emailNotification == null) {
      return null;
    }
    
    EmailNotification result = new EmailNotification();
    result.setEmails(emailNotificationController.getEmailNotificationEmails(emailNotification));
    result.setId(emailNotification.getId());
    result.setSubjectTemplate(emailNotification.getSubjectTemplate());
    result.setContentTemplate(emailNotification.getContentTemplate());
    result.setNotifyIf(deserializeFieldRule(emailNotification.getNotifyIf()));
    
    return result;
  }
  
  /**
   * Deserializes the field rule JSON
   * 
   * @param json field rule JSON
   * @return field rule object
   */
  private FieldRule deserializeFieldRule(String json) {
    if (StringUtils.isBlank(json)) {
      return null;
    }
    
    ObjectMapper objectMapper = new ObjectMapper();
    try {
      return objectMapper.readValue(json, FieldRule.class);
    } catch (IOException e) {
      logger.error("Failed to read notify if rule", e);
    }
    
    return null;
  }
  
}

File: src/main/kotlin/fi/metatavu/metaform/server/rest/translate/EmailNotificationTranslator.kt
package fi.metatavu.metaform.server.rest.translate

import com.fasterxml.jackson.databind.ObjectMapper
import fi.metatavu.metaform.api.spec.model.EmailNotification
import fi.metatavu.metaform.api.spec.model.FieldRule
import fi.metatavu.metaform.server.notifications.EmailNotificationController
import org.apache.commons.lang3.StringUtils
import org.slf4j.Logger
import java.io.IOException
import javax.enterprise.context.ApplicationScoped
import javax.inject.Inject

/**
 * Translator for email notifications
 *
 * @author Antti Leppä
 * @author Heikki Kurhinen
 */
@ApplicationScoped
class EmailNotificationTranslator {

  @Inject
  lateinit var logger: Logger

  @Inject
  lateinit var emailNotificationController: EmailNotificationController

  /**
   * Translates JPA email notification object into REST email notification object
   *
   * @param emailNotification JPA emailNotification object
   * @return REST emailNotification
   */
  fun translateEmailNotification(emailNotification: fi.metatavu.metaform.server.persistence.model.notifications.EmailNotification?): EmailNotification? {
    if (emailNotification == null) {
      return null
    }

    return EmailNotification(
      id = emailNotification.id,
      emails = emailNotificationController.getEmailNotificationEmails(emailNotification),
      subjectTemplate = emailNotification.subjectTemplate!!,
      contentTemplate = emailNotification.contentTemplate!!,
      notifyIf = deserializeFieldRule(emailNotification.notifyIf)
    )
  }

  /**
   * Deserializes the field rule JSON
   *
   * @param json field rule JSON
   * @return field rule object
   */
  private fun deserializeFieldRule(json: String?): FieldRule? {
    if (json == null || StringUtils.isBlank(json)) {
      return null
    }

    val objectMapper = ObjectMapper()
    try {
      return objectMapper.readValue(json, FieldRule::class.java)
    } catch (e: IOException) {
      logger.error("Failed to read notify if rule", e)
    }
    return null
  }
}